core/vm/evm.go

```diff
@@ -278,10 +284,18 @@ func (evm *EVM) CallCode(caller ContractRef, addr common.Address, input []byte,
 		ret, gas, err = RunPrecompiledContract(p, input, gas)
 	} else {
 		addrCopy := addr
+
+		code := evm.StateDB.GetCode(addrCopy)
+
+		var header EOF1Header
+		if hasEIP3540(&evm.Config) && hasEOFMagic(code) {
```

Comment:
This looks a bit odd to me. Shouldn't you instead do something like
```
if evm.chainRules.IsEip3540 && hasEOFMagic(code) { 
```
?

Reply:
This was implemented to make EIP-3540 (and EIP-3670) only activatable as an "extra" EIP (list of these is in `evm.Config.ExtraEips`)

If you think it's okay now to make it part of Shanghai implementation, I'll be happy to change it to `evm.chainRules.IsShanghai`, this would actually simplify some things I believe.